[WebCodecs] AudioDecoder and AudioData support

philn opened the pull request:
#### 1842861b125b97a693e2f4aa6bcd70e02b0b5bce
<pre>
[WebCodecs] AudioDecoder and AudioData support
<a href="https://bugs.webkit.org/show_bug.cgi?id=259874">https://bugs.webkit.org/show_bug.cgi?id=259874</a>
&lt;rdar://problem/113823411&gt;

Reviewed by Youenn Fablet.

This patch provides an initial implementation of the AudioDecoder and AudioData objects as defined
by the WebCodecs spec. Platform implementations are provided for the GStreamer ports.

The WebCodecsEnabled setting was renamed to WebCodecsVideoEnabled and a new WebCodecsAudioEnabled
setting was added, on by default for GStreamer ports.

Canonical link: <a href="https://commits.webkit.org/267278@main">https://commits.webkit.org/267278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbcc53118e19a683bf60e3d589de9ce36c5a6461

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16178 "19 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17943 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16606 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18708 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14069 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14641 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21478 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13931 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15058 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18037 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15416 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15400 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13057 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/16380 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14625 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4080 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3860 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18991 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/17546 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15222 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3872 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -0,0 +1,349 @@
+    ~GStreamerInternalAudioDecoder() = default;
+
+    void postTask(Function<void()>&& task) { m_postTaskCallback(WTFMove(task)); }
+    void decode(std::span<const uint8_t>, bool isKeyFrame, int64_t timestamp, std::optional<uint64_t> duration,  AudioDecoder::DecodeCallback&&);
```
There are 2 spaces after the last comma

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -0,0 +1,349 @@
+    });
+
+    GRefPtr<GstElement> element;
+    if (codecName.startsWith("pcm-"_s)) {
+        auto components = codecName.split('-');
+        if (components.size() != 2) {
+            GST_WARNING("Invalid LPCM codec string: %s", codecName.ascii().data());
+            return false;
+        }
```
In this branch there is no element to create.

philn replied:
As expected, this format needs no decoder.

calvaris replied:
Well, I guess that :) My question was more in case you needed an identify event in case the nullptr could be creating issues.

philn replied:
identity you mean? No that's not needed.

calvaris replied:
Yes, my brain-keyboard connection went bananas when writing that.

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -0,0 +1,349 @@
+
+        auto bufferSize = data.size();
+        auto bufferData = data.data();
+        auto codecData = adoptGRef(gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, bufferData, bufferSize, 0, bufferSize, new Vector<uint8_t>(WTFMove(data)), [](gpointer data) {
```
Check for span to be empty and create the vector here from the span

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    auto bufferSize = data.size();
+    auto bufferData = data.data();
+    auto buffer = adoptGRef(gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, bufferData, bufferSize, 0, bufferSize, new Vector<uint8_t>(WTFMove(data)), [](gpointer data) {
```
span vs vector comments above

calvaris commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -0,0 +1,236 @@
+
+RefPtr<PlatformRawAudioData> PlatformRawAudioData::create(std::span<const uint8_t>&& sourceData, AudioSampleFormat format, float sampleRate, int64_t timestamp, size_t numberOfFrames, size_t numberOfChannels)
+{
+    UNUSED_PARAM(numberOfFrames);
```
why don't you just remove the param name?

calvaris commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -0,0 +1,236 @@
+void PlatformRawAudioData::copyTo(std::span<uint8_t> destination, AudioSampleFormat format, size_t planeIndex, std::optional<size_t> frameOffset, std::optional<size_t> frameCount, unsigned long copyElementCount)
+{
+    UNUSED_PARAM(frameCount);
+    UNUSED_PARAM(copyElementCount);
```
ditto

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
```diff
@@ -185,11 +185,11 @@ void GStreamerElementHarness::start(GRefPtr<GstCaps>&& inputCaps)
 }
 
-void GStreamerElementHarness::pushStickyEvents(GRefPtr<GstCaps>&& inputCaps)
+void GStreamerElementHarness::pushStickyEvents(GRefPtr<GstCaps>&& inputCaps, std::optional<const GstSegment*> segment)
```
You're moving the optional into the other method and neither start nor pushStickyEvents take a &&. It is going to work but semantically it has little sense. You might want to reconsider this in one way or another

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:
```diff
@@ -85,7 +85,7 @@ class GStreamerElementHarness : public ThreadSafeRefCounted<GStreamerElementHarn
     ~GStreamerElementHarness();
 
-    void start(GRefPtr<GstCaps>&&);
+    void start(GRefPtr<GstCaps>&&, std::optional<const GstSegment*> = { });
```
Now that I see this, you might want to add && here and for the sticky event

calvaris reviewed (commented):
I added some comment regarding the GStreamer code. I'll let others review the rest.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioData.idl:
```diff
@@ -0,0 +1,67 @@
+// FIXME: Support Serializable and Transferable.
+[
+    Conditional=WEB_CODECS,
+    EnabledBySetting=WebCodecsEnabled,
```
Might be good to add its own WebCodecsAudioEnabled  setting, maybe we should also rename WebCodecsEnabled with WebCodecsVideoEnabled.

philn replied:
I wonder why? In Gecko it's a single setting for instance. In Chrome I couldn't find a flag about this.

youennf replied:
WebCodecsEnabled is true in some platforms but we might want to enable audio support only once it is fully implemented.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
```diff
@@ -0,0 +1,187 @@
+    if (!m_data.audioData)
+        return { };
+
+    return m_data.audioData->format();
```
Could be a one liner here and below probably.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
```diff
@@ -0,0 +1,187 @@
+    if (copyElementCount.hasException())
+        return copyElementCount.releaseException();
+
+    auto destFormat = options.format.value_or(*format());
```
In theory, there is nothing precluding having m_isDetached=false by m_data.audioData be null.

philn replied:
I'm sorry I don't understand this comment :)

youennf replied:
How is it guaranteed that format() is not std::nullopt here?

philn replied:
Ah right, I understand now, thanks

philn replied:
I refactored `computeCopyElementCount()` to raise an exception earlier if internal data is null, so I think we can keep this as-is.

youennf replied:
Maybe computeCopyElementCount could also return the destination format as well.
computeCopyElementCount uses method that repeatedly checks internal data is not null, but I guess it is fine as is.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioInternalData.h:
```diff
@@ -0,0 +1,44 @@
+struct WebCodecsAudioInternalData {
+    size_t memoryCost() const { return audioData ? audioData->memoryCost() : 0; }
+
+    RefPtr<PlatformRawAudioData> audioData;
```
I wonder whether this struct is needed if audioData is planned to remain the sole member.
Do we expect to add more fields in the future?

philn replied:
Hum I'm not sure... I used the same pattern as for VideoFrame... Maybe we can add a FIXME and re-assess later on?

youennf replied:
OK, that is fine as is.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -0,0 +1,132 @@
+
+    auto totalSamples = init.numberOfFrames * init.numberOfChannels;
+    auto bytesPerSample = computeBytesPerSample(init.format);
+    auto totalSize = totalSamples * bytesPerSample;
```
We can have some multiplication overflows in these two multiplications, which might break the comparison between dataSize and totalSize.
Let's add some tests and make use of safeMultiply.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -0,0 +1,132 @@
+// https://www.w3.org/TR/webcodecs/#compute-copy-element-count
+ExceptionOr<size_t> computeCopyElementCount(const WebCodecsAudioData& data, const WebCodecsAudioData::CopyToOptions& options)
+{
+    auto destFormat = options.format.value_or(*data.format());
```
How do we know data.format() is not null opt here?

philn replied:
This is ambiguous indeed, I'll refactor this a bit :)

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    if (isInterleaved)
+        copyFrameCount *= data.numberOfChannels();
```
Ditto for these multiplications, maybe we should use safe operations and throw if it overflows.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
```diff
@@ -0,0 +1,44 @@
+
+bool isValidAudioDataInit(const WebCodecsAudioData::Init&);
+
+bool isAudioSampleFormatInterleaved(const AudioSampleFormat&);
```
We could forward declare this one and remove the corresponding include

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -0,0 +1,323 @@
+    if (!config.codec.startsWith("mp4a."_s) && config.codec != "mp3"_s && config.codec != "opus"_s && config.codec != "alaw"_s && config.codec != "ulaw"_s && config.codec != "flac"_s && config.codec != "vorbis"_s && !config.codec.startsWith("pcm-"_s))
+        return false;
+
+    return true;
```
We could rewrite it as:
return config.codec.startsWith() || ...

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -0,0 +1,323 @@
+ExceptionOr<void> WebCodecsAudioDecoder::configure(ScriptExecutionContext& context, WebCodecsAudioDecoderConfig&& config)
+{
+    if (!isValidDecoderConfig(config, context.settingsValues()))
+        return Exception { TypeError, "Config is not valid"_s };
```
This is fine as is, but the spec is moving now to close the decoder with NotSupportedError instead.
I'll do the change for video decoder/encoder this week.

youennf commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 51;
```
Not needed

youennf commented on Source/WebCore/platform/AudioDecoder.cpp:
```diff
@@ -0,0 +1,78 @@
+
+void AudioDecoder::setCreatorCallback(CreatorFunction&& function)
+{
+    s_customCreator = WTFMove(function);
```
I wonder whether we will need a custom creator, I am not sure we will want to do encoding/decoding in GPU process.

philn replied:
I think that whenever the GStreamer ports migrate to GPUProcess this would be needed... I can drop if for now though.

youennf replied:
Let's drop it for now then.

youennf reviewed (commented):
Looks good to me overall. Some questions/suggestions below.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
```diff
@@ -0,0 +1,169 @@
+    auto destFormat = options.format.value_or(*format());
+    auto bytesPerSample = computeBytesPerSample(destFormat);
+    auto maxCopyElementCount = copyElementCount.releaseReturnValue();
+    auto allocationSize = maxCopyElementCount * bytesPerSample;
```
Probably safer to do safe multiplications as well here.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
```diff
@@ -0,0 +1,169 @@
+
+    auto clone = adoptRef(*new WebCodecsAudioData(context, WebCodecsAudioInternalData { m_data }));
+
+    clone->m_isDetached = m_isDetached;
```
Do we need this given we reject when AudioData is detached?

philn replied:
No need indeed.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    auto destFormat = platformData->format();
+    if (options.format)
+        destFormat = *options.format;
```
`auto destFormat = options.format.value_or(platformData->format());`

philn committed:
[WebCodecs] AudioDecoder and AudioData support
https://bugs.webkit.org/show_bug.cgi?id=259874
<rdar://problem/113823411>

Reviewed by Youenn Fablet.

This patch provides an initial implementation of the AudioDecoder and AudioData objects as defined
by the WebCodecs spec. Platform implementations are provided for the GStreamer ports.

The WebCodecsEnabled setting was renamed to WebCodecsVideoEnabled and a new WebCodecsAudioEnabled
setting was added, on by default for GStreamer ports.

Canonical link: https://commits.webkit.org/267278@main
